stl/inc/type_traits

```diff
@@ -366,6 +366,15 @@ struct is_enum : bool_constant<__is_enum(_Ty)> {}; // determine whether _Ty is a
 template <class _Ty>
 _INLINE_VAR constexpr bool is_enum_v = __is_enum(_Ty);
 
+#if _HAS_CXX23
+// STRUCT TEMPLATE is_scoped_enum
+template <class _Ty>
+inline constexpr bool is_scoped_enum_v = conjunction_v<is_enum<_Ty>, negation<is_convertible<_Ty, int>>>;

```

Comment:
```cpp
enum E  {
   e = std::is_scoped_enum_v<E>
};
static_assert(!e);
```

Reply:
That's a ~~diabolical~~ ~~brilliant~~ brilliantly diabolical example. I think this means we need a compiler builtin.

The Standardese is WG21-N4885 [dcl.enum]/6 "An enumeration whose underlying type is not fixed is an incomplete type until the closing `}` of its *enum-specifier*, at which point it becomes a complete type." and [meta.rqmts]/5 "Unless otherwise specified, an incomplete type may be used to instantiate a template specified in 20.15." There are no additional preconditions for `is_scoped_enum`, so the incomplete `enum E` is permitted.

Reply:
So I'm a little confused. Is that not supposed to compile?
```cpp
enum E  {
   e1 = std::is_scoped_enum_v<E>,
   e2 = std::is_enum_v<E>
};
static_assert(!e1);
static_assert(e2);
```
This compiles just fine. If `is_scoped_enum_v` isn't supposed to work here then why does `is_enum_v`?

Reply:
IIUC
```
enum E  {
   a = std::is_convertible<E, int>, // IFNDR because E is incomplete
   e = std::is_scoped_enum_v<E> // should be ok; value is false
};
```

But I *think* it's valid to rely on the implementation-specific knowledge that `std::is_convertible_v` works in such case.

Reply:
Ah, that makes sense. Thanks.

Reply:
Fascinating. MSVC on godbolt apparently even accepts (in violation of http://eel.is/c++draft/meta.trans.other#tab:meta.trans.other-row-12-column-2-sentence-2)

```cpp 
enum E {
    e = std::underlying_type_t<E>(42)
};
```

I wonder what it thinks `E`'s underlying type is...

Reply:
The current implementation of `is_scoped_enum` does compile with clang as well btw. And by the looks of it, from my untrained eyes, it's similar  to the libc++'s implementation.

Reply:
libc++'s implementation is even more broken - it actually uses the underlying type, which diagnoses incomplete enums in clang. Clang's handling of incomplete enums seems to have regressed between 9 and 10.

A testament to how rare such types are in the wild, I suppose.

Reply:
MSVC currently believes (noncomformingly) that enumeration types are always complete, and that the underlying type is `int` until specified otherwise. We allow `enum meow` as a forward declaration, even in strict mode: https://godbolt.org/z/MoMnn4zxz. IIRC the Windows SDK needed this to work as of two years ago, no idea if it's been fixed yet.

Reply:
That second example is quite the eyebrow raiser...

Reply:
> I think we can proceed with the current implementation then. If we discover that the issue is worse than previously thought, we can always request a compiler builtin and replace the implementation.

I'm inclined to request intrinsics now so the library can be prepared for MSVC's enumeration bugs to be fixed. When some compiler dev goes to audit what things break with the bug fixed, I'd rather we don't show up on that list.

Reply:
> I think we can proceed with the current implementation then. If we discover that the issue is worse than previously thought, we can always request a compiler builtin and replace the implementation.

FWIW, CWG-2516 has been accepted as a DR, so a scoped enum type is always complete when its name is declared, and henc no intrinsic is needed for `is_scoped_enum`. But we may still have to replace the current implementation once MSVC's bug (or non-conforming extension) is fixed.

Reply:
I noticed that MSVC has implemented [`/Zc:enumTypes` (Enable enum type deduction)](https://learn.microsoft.com/en-us/cpp/build/reference/zc-enumtypes?view=msvc-170), but it seems that despite this option, MSVC still believes that enumeration types are always complete, and that the underlying type is `int` before the closing `}`.

```c++
#include <type_traits>

template<class T>
constexpr std::underlying_type_t<T> f() {
  T t;
  return 1;
}

enum E {
  a = f<E>(),  // should be ill-formed
  b = sizeof(E),  // should be ill-formed
  c = 0x1'0000'0000,
};

static_assert(b == sizeof(E)); // fires under /Zc:enumTypes
```